### ISSUE TITLE
Create nodes with alias and menu item

### ISSUE THREAD

**Issue by attiks** [jhedstrom/drupalextension#497]
See #295 

I needed a way to create an url alias and a menu item for some nodes, I used the following code, it might be that it can be optimised into one method.

```php
<?php

use Drupal\DrupalExtension\Context\RawDrupalContext;
use Behat\Behat\Context\SnippetAcceptingContext;
use Behat\Gherkin\Node\PyStringNode;
use Behat\Gherkin\Node\TableNode;
use Behat\Behat\Tester\Exception\PendingException;
use Behat\Mink\Driver\Selenium2Driver;
use Behat\Mink\Exception\ElementNotFoundException;
use Behat\Behat\Hook\Scope\AfterStepScope;
use Behat\Behat\Hook\Scope\BeforeStepScope;

/**
 * Defines application features from the specific context.
 */
class FeatureContext extends RawDrupalContext implements SnippetAcceptingContext {

  /**
   * @Given I am viewing a :arg1 content with the alias :arg2:
   */
  public function iAmViewingAContentWithTheAlias($type, $alias, TableNode $fields) {
    $node = (object) array(
      'type' => $type,
    );

    foreach ($fields->getRowsHash() as $field => $value) {
      $node->{$field} = $value;
    }

    $saved = $this->nodeCreate($node);

    // Save path.
    $path = array(
      'source' => 'node/' . $saved->nid,
      'alias' => $alias,
    );
    path_save($path);

    // Set internal browser on the node.
    $this->getSession()->visit($this->locatePath('/node/' . $saved->nid));
  }

  /**
   * @Given I am viewing a :arg1 content with the alias :arg2 and menu :menu:
   */
  public function iAmViewingAContentWithTheAliasAndMenu($type, $alias, $menu, TableNode $fields) {
    $node = (object) array(
      'type' => $type,
    );

    foreach ($fields->getRowsHash() as $field => $value) {
      $node->{$field} = $value;
    }

    $saved = $this->nodeCreate($node);

    // Save path.
    $path = array(
      'source' => 'node/' . $saved->nid,
      'alias' => $alias,
    );
    path_save($path);

    // Save menu.
    if (!empty($menu)) {
      list($menu_name, $link_title) = explode(':', $menu);
      $item = array(
        'menu_name' => $menu_name,
        'link_path' => drupal_get_normal_path($alias),
        'link_title' => $link_title,
        'weight' => 0,
        'customized' => 1,
      );
      menu_link_save($item);
      menu_cache_clear_all();
    }

    // Set internal browser on the node.
    $this->getSession()->visit($this->locatePath('/node/' . $saved->nid));
  }
}
```

**Comment by jhedstrom** [comment]
This is a great example, thanks for posting! Hopefully it will help others in the future. Closing out for now since there isn't much more to be done here.

**Comment by xurizaemon** [comment]
This allows use of the alias in `Given :type content:` tables:

In FeatureContext:

```php
  /**
   * Test alias creation.
   *
   * @afterNodeCreate
   */
  public function nodePostSave(AfterNodeCreateScope $event) {
    $entity = $event->getEntity();
    if (isset($entity->path_alias) && isset($entity->nid)) {
      $path_alias = \Drupal\path_alias\Entity\PathAlias::create([
        'path' => '/node/' . $entity->nid,
        'alias' => $entity->path_alias,
      ]);
      $path_alias->save();
    }
  }
```

Usage:

```feature
  Scenario: Testing page creation with aliases
    Given "page" content:
      | title     | body      | published | | path_alias |
      | Test page | Some body | 1         | | /ex-4-path |
```
Happy to PR this - it depends on Drupal 8.8, and the path creation code in OP above will work for Drupal 7, so we'll need to make it do this based on active driver.

**Comment by pfrenssen** [comment]
Yeah DrupalExtension is supposed to be agnostic of the version of Drupal that is used, so it is always a good idea to build on top of methods that are provided in DrupalDriver.

**Comment by phuang07** [comment]
> This allows use of the alias in `Given :type content:` tables:
> 
> In FeatureContext:
> 
> ```
>   /**
>    * Create aliases based on provided path_alias value.
>    *
>    * @afterNodeCreate
>    */
>   public function nodePathAliasPostSave(AfterNodeCreateScope $event) {
>     $entity = $event->getEntity();
>     if (isset($entity->path_alias) && isset($entity->nid)) {
>       $path_alias = \Drupal\path_alias\Entity\PathAlias::create([
>         'path' => '/node/' . $entity->nid,
>         'alias' => $entity->path_alias,
>       ]);
>       $path_alias->save();
>     }
>   }
> ```
> 
> Usage:
> 
> ```
>   Scenario: Testing page creation with aliases
>     Given "page" content:
>       | title     | body      | published | | path_alias |
>       | Test page | Some body | 1         | | /ex-4-path |
> ```
> 
> Happy to PR this if it'd be a good addition.
> 
> Notes:
> 
> * Implementation depends on Drupal 8.8 core here, and D7 in OP.
> * May need to live in the driver as a result?
> * Should test whether the appropriate module is enabled before calling
> * If the module's not enabled, is it still OK to have `use Drupal\path_alias\Entity\PathAlias` in the driver?

I am getting the following error:
" Argument 1 passed to FeatureContext::nodePathAliasPostSave() must be an instance of AfterNodeCreateScope, instance of Drupal\DrupalExtension\Hook\Scope\AfterNodeCreateScope given "

Any clue how this can be fixed? I am running on Drupal 8.9.x. Thanks.

**Comment by xurizaemon** [comment]
> I am getting the following error:
> " Argument 1 passed to FeatureContext::nodePathAliasPostSave() must be an instance of AfterNodeCreateScope, instance of Drupal\DrupalExtension\Hook\Scope\AfterNodeCreateScope given "

Hi @phuang07, I've added the `use` statement omitted before to my comment above. Add this to the top of your FeatureContext.
